File: src-tauri/src/commands/io.rs
//! Session I/O Tauri commands.
//!
//! Commands for sending prompts to sessions.

use crate::commands::{write_frames_to, MediaItem};
use crate::session::{self, SessionMap};
use crate::tlv;

use tauri::State;

#[tauri::command]
pub async fn alayacore_send_prompt(
    session_id: String,
    text: String,
    media: Vec<MediaItem>,
    sessions: State<'_, SessionMap>,
) -> Result<(), String> {
    // Build the whole message (media + text + UE flush) first, then write
    // it under ONE stdin lock — and crucially WITHOUT holding the
    // SessionMap lock. A prompt carries media data URIs (megabytes) on a
    // ~64 KiB OS pipe: if this one alayacore stops draining stdin, the
    // write blocks, and holding the global map lock across it used to
    // freeze every other session's commands (Go's SendPrompt only ever
    // takes the per-session stdin lock).
    let mut frames: Vec<(String, String)> = Vec::with_capacity(media.len() + 2);
    for item in &media {
        let tag = match item.media_type.as_str() {
            "image" => tlv::TAG_USER_IMAGE,
            "audio" => tlv::TAG_USER_AUDIO,
            "video" => tlv::TAG_USER_VIDEO,
            "document" => tlv::TAG_USER_DOC,
            _ => return Err(format!("Unknown media type: {}", item.media_type)),
        };
        frames.push((tag.to_string(), item.uri.clone()));
    }
    if !text.is_empty() {
        frames.push((tlv::TAG_USER_TEXT.to_string(), text));
    }
    frames.push((tlv::TAG_USER_END.to_string(), String::new()));

    let refs = session::refs(sessions.inner(), &session_id).await?;
    write_frames_to(&refs, &session_id, &frames).await
}

File: src-go/internal/server/handlers/io.go
package handlers

import (
	"fmt"
	"net/http"

	"alayaface/src-go/internal/tlv"
)

// MediaItem mirrors the Rust MediaItem (serde keeps snake_case).
type MediaItem struct {
	MediaType string `json:"media_type"`
	URI       string `json:"uri"`
}

// SendPrompt sends a user message (with optional media) to a session,
// ending with a UE flush frame. Port of commands/io.rs alayacore_send_prompt.
func SendPrompt(h *Handler, w http.ResponseWriter, r *http.Request) error {
	var args struct {
		SessionID string      `json:"sessionId"`
		Text      string      `json:"text"`
		Media     []MediaItem `json:"media"`
	}
	if err := decodeArgs(r, &args); err != nil {
		return err
	}

	s, err := h.Sessions.Get(args.SessionID)
	if err != nil {
		return err
	}
	if !s.Connected() {
		return fmt.Errorf("Session is disconnected")
	}

	// Build the whole message (media + text + UE) and write it under a
	// single stdin lock so a concurrent CI command cannot interleave.
	frames := make([]tlv.Frame, 0, len(args.Media)+2)
	for _, item := range args.Media {
		var tag string
		switch item.MediaType {
		case "image":
			tag = tlv.TagUserImage
		case "audio":
			tag = tlv.TagUserAudio
		case "video":
			tag = tlv.TagUserVideo
		case "document":
			tag = tlv.TagUserDoc
		default:
			return fmt.Errorf("Unknown media type: %s", item.MediaType)
		}
		frames = append(frames, tlv.Frame{Tag: tag, Value: item.URI})
	}
	if args.Text != "" {
		frames = append(frames, tlv.Frame{Tag: tlv.TagUserText, Value: args.Text})
	}
	frames = append(frames, tlv.Frame{Tag: tlv.TagUserEnd, Value: ""})
	if err := s.WriteFrames(frames); err != nil {
		return err
	}
	return writeResult(w, nil)
}
